crates/aptos/src/account/key_rotation.rs

```diff
@@ -286,41 +283,66 @@ impl CliCommand<AccountAddress> for LookupAddress {
         // TODO: This command needs to support multi-ed25519
-        let address_key = AuthenticationKey::ed25519(&self.public_key()?).derived_address();
-        match rest_client
-            .get_table_item_bcs(
-                table_handle,
-                "address",
-                "address",
-                address_key.to_hex_literal(),
-            )
-            .await
-        {
-            Ok(inner) => Ok(inner.into_inner()),
-            Err(RestError::Api(AptosErrorResponse {
+        let address_key = account_address_from_public_key(&self.public_key()?);
+        Ok(lookup_address(&rest_client, address_key, true).await?)
+    }
+}
+
+pub async fn lookup_address(
+    rest_client: &Client,
+    address_key: AccountAddress,
+    must_exist: bool,
+) -> Result<AccountAddress, RestError> {
+    let originating_resource: OriginatingResource = rest_client
+        .get_account_resource_bcs(CORE_CODE_ADDRESS, "0x1::account::OriginatingAddress")
+        .await?
+        .into_inner();
+
+    let table_handle = originating_resource.address_map.handle;
+
+    // The derived address that can be used to look up the original address
+    match rest_client
+        .get_table_item_bcs(
+            table_handle,
+            "address",
+            "address",
+            address_key.to_hex_literal(),
+        )
+        .await
+    {
+        Ok(inner) => Ok(inner.into_inner()),
+        Err(RestError::Api(AptosErrorResponse {
+            error:
+                AptosError {
+                    error_code: AptosErrorCode::TableItemNotFound,
+                    ..
+                },
+            ..
+        })) => {
+            // If the table item wasn't found, let's at least check if the account exists
+            // It won't be in the table if it wasn't rotated, then return the derived account address
+            let result = rest_client.get_account_bcs(address_key).await;
+
+            if let Err(RestError::Api(AptosErrorResponse {
                 error:
                     AptosError {
-                        error_code: AptosErrorCode::TableItemNotFound,
+                        error_code: AptosErrorCode::AccountNotFound { .. },
                         ..
                     },
                 ..
-            })) => {
-                // If the table item wasn't found, let's at least check if the account exists
-                // It won't be in the table if it wasn't rotated, then return the derived account address
-                rest_client.get_account_bcs(address_key).await?;
-                Ok(address_key)
-            },
-            Err(err) => Err(err)?,
-        }
+            })) = result
+            {
+                if must_exist {
+                    return Err(result.unwrap_err());
+                }
+            } else if let Err(err) = result {
+                return Err(err);
+            }
+
+            Ok(address_key)
```

Comment:
I find this awkward, can we just use a match statement, I feel it is more readable:
```
match rest_client.get_account_bcs(address_key).await {
    Ok(_) => Ok(address_key),
    Err(RestError::Api(AptosErrorResponse { etc })) => etc,
    Err(e) => Err(e),
}
```

Makes the control flow simpler, no explicit early returns, etc.

Reply:
I did this awkward flow, because as you can tell, it had duplicated code from this one special case / the result isn't the same

Reply:
If you transform the result type it should be possible right?

Reply:
Anyway this is a high pri fix so don't worry about it if it's too tricky, we can come back to it later.